python/sglang/srt/layers/sampler.py

```diff
@@ -50,6 +50,20 @@ def forward(
         else:
             # Post process logits
             logits.div_(sampling_info.temperatures)
+
+            if any(sampling_info.top_n_sigmas > 0):
+                max_logit = torch.max(logits, dim=-1, keepdim=True).values
+                sigma = torch.std(logits, dim=-1, keepdim=True)
+                # Create mask and enable only for the requests that have top_n_sigma > 0
+                mask = (sampling_info.top_n_sigmas.view(-1, 1) <= 0) | (
+                    logits >= max_logit - sampling_info.top_n_sigmas.view(-1, 1) * sigma
+                )
+
+                # Apply mask
+                logits = torch.where(
+                    mask, logits, torch.tensor(float("-inf")).to(logits.device)
+                )
```

Comment:
To prevent this function from being too long, wrap this as a function and call it here.